helix-core/src/test/java/org/apache/helix/integration/paticipant/TestStateTransitionAppFailureHandling.java

```diff
@@ -0,0 +1,203 @@
+            accessor.getChildNames(keyBuilder.messages(instanceName)).stream()
+                .map(childName -> keyBuilder.message(instanceName, childName))
+                .collect(Collectors.toList()), true);
+        for (Message message : messageList) {
+          if (message.getMsgState() != Message.MessageState.UNPROCESSABLE) {
+            return false;
+          }
+        }
+        partitionMessageMap.put(instanceName, messageList);
+        totalMessageCount += messageList.size();
+      }
+      return totalMessageCount == _PARTITIONS * REPLICAS;
+    }, TestHelper.WAIT_DURATION));
+
+    // Verify that the correct numbers of retry has been done on each node.
+    for (String instanceName : partitionMessageMap.keySet()) {
+      List<Message> instanceMessages = partitionMessageMap.get(instanceName);
+      for (Message message : instanceMessages) {
+        Assert.assertTrue(message.getRetryCount() <= 0);
+        Assert.assertEquals(message.getMsgState(), Message.MessageState.UNPROCESSABLE);
+      }
+      // Check if the factory has tried enough times before fail the message.
+      Assert.assertEquals(retryCountUntilSucceed - retryFactoryMap.get(instanceName)
+          .getRemainingRetryCountUntilSucceed(), instanceMessages.size()
+          * MessageGenerationPhase.DEFAULT_STATE_TRANSITION_MESSAGE_RETRY_COUNT);
+    }
+
+    // Verify that the partition is not initialized.
+    for (int i = 0; i < NODE_NR; i++) {
+      String instanceName = _participants[i].getInstanceName();
+      String sessionId = _participants[i].getSessionId();
+      List<CurrentState> currentStates = accessor.getProperty(
+          accessor.getChildNames(keyBuilder.currentStates(instanceName, sessionId)).stream()
+              .map(childName -> keyBuilder.currentState(instanceName, sessionId, childName))
+              .collect(Collectors.toList()), true);
+      for (CurrentState currentState : currentStates) {
+        Assert.assertTrue(currentState.getPartitionStateMap().isEmpty());
+      }
+    }
+
+    _gSetupTool.dropResourceFromCluster(CLUSTER_NAME, TEST_DB);
+  }
+
+  @Test(dependsOnMethods = "testSTHandlerInitFailureRetry")
+  public void testSTHandlerInitFailureRetrySucceed() {
+    // Make the mock StateModelFactory return handler before last retry. So it will successfully
+    // finish handler initialization.
+    int retryCountUntilSucceed =
+        MessageGenerationPhase.DEFAULT_STATE_TRANSITION_MESSAGE_RETRY_COUNT - 1;
+    Map<String, RetryStateModelFactory> retryFactoryMap = resetParticipants(retryCountUntilSucceed);
+
+    _gSetupTool.addResourceToCluster(CLUSTER_NAME, TEST_DB, _PARTITIONS, STATE_MODEL);
+    _gSetupTool.rebalanceStorageCluster(CLUSTER_NAME, TEST_DB, REPLICAS);
+
+    HelixDataAccessor accessor = _controller.getHelixDataAccessor();
+    PropertyKey.Builder keyBuilder = accessor.keyBuilder();
+
+    // Verify and wait until all messages have been retried and failed.
```

Comment:
You mean retried and succeeded finally, right?

Reply:
Fixed.